types/sorted_cid_set.go

```diff
@@ -5,188 +5,173 @@ import (
-	cbor.RegisterCborType(atlas.BuildEntry(SortedCidSet{}).Transform().
+	// A TipSetKey serializes as a sorted array of CIDs.
+	// Deserialization will sort the CIDs, if they're not already.
+	cbor.RegisterCborType(atlas.BuildEntry(TipSetKey{}).Transform().
 		TransformMarshal(atlas.MakeMarshalTransformFunc(
-			func(s SortedCidSet) ([]cid.Cid, error) {
-				return s.s, nil
+			func(s TipSetKey) ([]cid.Cid, error) {
+				return s.cids, nil
 			})).
 		TransformUnmarshal(atlas.MakeUnmarshalTransformFunc(
-			func(s []cid.Cid) (SortedCidSet, error) {
-				for i := 0; i < len(s)-1; i++ {
-					// Note that this will also catch duplicates.
-					if !cidLess(s[i], s[i+1]) {
-						return SortedCidSet{}, fmt.Errorf(
-							"invalid serialization of SortedCidSet - %s not less than %s", s[i].String(), s[i+1].String())
-					}
-				}
-				return SortedCidSet{s: s}, nil
+			func(cids []cid.Cid) (TipSetKey, error) {
+				return NewTipSetKeyFromUnique(cids...)
 			})).
 		Complete())
 }
 
-// SortedCidSet is a set of Cids that is maintained sorted. The externally visible effect as
-// compared to cid.Set is that iteration is cheap and always in-order.
-// Sort order is lexicographic ascending, by serialization of the cid.
-// TODO: This should probably go into go-cid package - see https://github.com/ipfs/go-cid/issues/45.
-type SortedCidSet struct {
-	s []cid.Cid // should be maintained sorted
+// TipSetKey is an immutable set of CIDs forming a unique key for a TipSet.
+// Equal keys will have equivalent iteration order, but note that the CIDs are *not* maintained in
+// the same order as the canonical iteration order of blocks in a tipset (which is by ticket).
+// TipSetKey is a lightweight value type; passing by pointer is usually unnecessary.
+type TipSetKey struct {
+	// The slice is wrapped in a struct to enforce immutability.
+	cids []cid.Cid
 }
 
-// NewSortedCidSet returns a SortedCidSet with the specified items.
-func NewSortedCidSet(ids ...cid.Cid) (res SortedCidSet) {
-	for _, id := range ids {
-		res.Add(id)
+// NewTipSetKey initialises a new TipSetKey.
+// Duplicate CIDs are silently ignored.
+func NewTipSetKey(ids ...cid.Cid) TipSetKey {
+	if len(ids) == 0 {
+		// Empty set is canonically represented by a nil slice rather than zero-length slice
+		// so that a zero-value exactly matches an empty one.
+		return TipSetKey{}
 	}
-	return
-}
 
-// Add adds a cid to the set. Returns true if the item was added (didn't already exist), false
-// otherwise.
-func (s *SortedCidSet) Add(id cid.Cid) bool {
-	idx := s.search(id)
-	if idx < len(s.s) && s.s[idx].Equals(id) {
-		return false
+	cids := make([]cid.Cid, len(ids))
+	copy(cids, ids)
+	sort.Slice(cids, func(i, j int) bool {
+		return cidLess(cids[i], cids[j])
+	})
+
+	if len(cids) >= 2 {
+		// Uniq-ify the sorted array.
+		this, next := 0, 1
+		for next < len(cids) {
+			if cids[next] != cids[this] {
+				this++
+				cids[this] = cids[next]
+			}
+			next++
+		}
+		return TipSetKey{cids[:this+1]}
```

Comment:
If you keep it a descriptive comment (maybe a link describing this technique if it's standard) will help

Reply:
I wouldn't be convinced by any pencil and paper work, so I wrote tests :-)

It's not a particularly standard/common technique (or problem) AFAIK. You're right, if I were doing it again using a second slice would also have been faster to write. I didn't expect it to come out this "clever". Since it's now here and tested, I'm deciding to leave it in. I added an explanatory comment.